File: JAVA/vendingmachine/ReadyState.java
package vendingmachine;

public class ReadyState implements VendingMachineState {
    private final VendingMachine vendingMachine;
    public ReadyState(VendingMachine vendingMachine) {
        this.vendingMachine = vendingMachine;
    }

    @Override
    public void selectProduct(Product product) {
        System.out.println("Product already selected: " + vendingMachine.getSelectedProduct().getName() + ". Please make a payment.");
    }

    @Override
    public void insertCoin(Coin coin) {
        vendingMachine.addCoin(coin);
        System.out.println("Coin inserted: " + coin.getValue());
        checkPaymentStatus();
    }

    @Override
    public void insertNote(Note note) {
        vendingMachine.addNote(note);
        System.out.println("Note inserted: " + note.getValue());
        checkPaymentStatus();
    }

    @Override
    public void dispenseProduct() {
        System.out.println("Please make a complete payment first.");
    }

    @Override
    public void returnChange() {
        System.out.println("Please make a payment first.");
    }

    private void checkPaymentStatus() {
        if(vendingMachine.getTotalPayment() >= vendingMachine.getSelectedProduct().getPrice()) {
            vendingMachine.setState(vendingMachine.getDispensingState());
        }
    }    
}

File: CPP/vendingmachine/src/ReadyState.cpp
#include "ReadyState.h"
#include "VendingMachine.h"
#include <iostream>

void ReadyState::selectProduct(VendingMachine *vendingMachine,
                               const Product &product)
{
  std::cout << "Product already selected. Please make payment." << std::endl;
}

void ReadyState::insertCoin(VendingMachine *vendingMachine, Coin coin)
{
  vendingMachine->addCoin(coin);
  std::cout << "Coin inserted." << std::endl;
  checkPaymentStatus(vendingMachine);
}

void ReadyState::insertNote(VendingMachine *vendingMachine, Note note)
{
  vendingMachine->addNote(note);
  std::cout << "Note inserted." << std::endl;
  checkPaymentStatus(vendingMachine);
}

void ReadyState::dispenseProduct(VendingMachine *vendingMachine)
{
  std::cout << "Please make payment first." << std::endl;
}

void ReadyState::returnChange(VendingMachine *vendingMachine)
{
  double change = vendingMachine->totalPayment -
                  vendingMachine->selectedProduct->getPrice();
  if (change > 0)
  {
    std::cout << "Change returned: " << change << "Rs." << std::endl;
    vendingMachine->resetPayment();
  }
  else
  {
    std::cout << "No change to return." << std::endl;
  }
  vendingMachine->setState(&vendingMachine->idleState);
}

void ReadyState::checkPaymentStatus(VendingMachine *vendingMachine)
{
  if (vendingMachine->totalPayment >=
      vendingMachine->selectedProduct->getPrice())
  {
    vendingMachine->setState(&vendingMachine->dispenseState);
  }
}
